### PR TITLE
On gem yank reserve namespace for 100 days if gem is older than a month

### PR DESCRIPTION
**Pull request by sonalkr132**
Addresses: #1226 

Approach taken [from](https://github.com/rubygems/rubygems.org/issues/1226#issuecomment-200831744):
> When a gem author yanks all versions of a gem, one of three things will happen:
i. If the gem is less than a month old, we will open the namespace and allow anyone to claim the gem name. (This is what happens for all gems now.)
ii. If the gem is more than a month old, we will lock the namespace and prevent anyone from pushing any versions for 100 days. We will display a message on the website informing users that this gem name might be available, but they will need to contact support for assistance. We can then verify on a case by case basis.
iii. After 100 days, this namespace will become available. This should be enough time for the community to realize something has happened and we can step in and fix it. Or it is not a problem and someone else can have the gem name.

Reserved gem page:
screenshot from 2016-07-07 20-18-09

### CODE REVIEW

File: app/models/rubygem.rb [hunk 1]
```diff
@@ -177,7 +177,9 @@ def to_param
   end
 
   def pushable?
-    new_record? || versions.indexed.count.zero?
+    return true if new_record?
+    return false if versions.indexed.count.nonzero?
+    updated_at < 100.days.ago || created_at > 30.days.ago
```
Comment (simi) [inline app/models/rubygem.rb]: What about to move this condition to `protected?` method or similar?

You can use simple readable condition after
```ruby
new_record? || versions.indexed.none? || !protected?
```
or similar.

File: app/views/rubygems/show.html.erb [hunk 2]
```diff
@@ -12,7 +12,11 @@
 
 <% if @rubygem.public_versions.count.zero? %>
   <div class="t-body">
-    <p><%= t '.not_hosted_notice' %></p>
+    <% if @rubygem.pushable? %>
+      <p><%= t '.not_hosted_notice' %></p>
+    <% else %>
+      <p><%= (t '.reserved_namespace', contact: link_to('Contact', 'http://help.rubygems.org/')).html_safe %></p>
```
Comment (simi) [inline app/views/rubygems/show.html.erb]: *just nitpick*

Why not to use classic approach?

```ruby
t('.reserved_namespace', contact: link_to('Contact', 'http://help.rubygems.org/')).html_safe
```

File: app/models/rubygem.rb (hunk 1 above)
Comment (sonalkr132) [inline app/models/rubygem.rb]: Although, I agree that:
```Ruby
return true if new_record?
return false if versions.indexed.count.nonzero?
updated_at < 100.days.ago || created_at > 30.days.ago
```
is not very elegant, using:
```Ruby
new_record? || versions.indexed.none? || !protected?
```
would make every rubygem older than a month and updated in last 100 days pushable. `protected?` function could look like:
```Ruby
# gem with no indexed versions and updated(yanked) in last 100 days and older than a month
versions.indexed.none? && updated_at > 100.days.ago && created_at < 30.days.ago
```
I didn't want to write a long boolean query to begin with, hence those guard statements.
May be there could be a functions which just checks for date range?

Let me know what do you think :smiley:

File: app/views/rubygems/show.html.erb (hunk 2 above)
Comment (sonalkr132) [inline app/views/rubygems/show.html.erb]: > just nitpick

Appreciate it.
tumblr_inline_mus80klotk1r4zbjb

File: app/models/rubygem.rb (hunk 1 above)
Comment (simi) [inline app/models/rubygem.rb]: Damn, I missed that. sorry :(

What about something like this (this passes your tests).

```ruby
  def pushable?
    new_record? || (versions.indexed.none? && not_protected?)
  end

  def not_protected?
    updated_at < 100.days.ago || created_at > 30.days.ago
  end
```

But maybe that's bad idea. Who knows.

**Comment by indirect** [comment]
This is a good solution, thanks! In follow up PRs, I would like to add download-based limits and also add this error message for all of the gems on the stdlib blacklist. 👍

**Comment by sonalkr132** [comment]
>  I would like to add download-based limits

I didn't realize that we would have both kind of limits. [The discussion](https://github.com/rubygems/rubygems.org/issues/1226#issuecomment-200831744) seems to suggest that we will use either one of two proposals. Can you please explain why limiting on basis of **either** `date` **or**  `download` is not enough?

**Comment by indirect** [comment]
@sonalkr132 I think it is valuable and important to protect gems that are either old (and expected to stay the same) or heavily used (and so removing them will break a lot of things).

I am not sure if 15,000 is the right limit (and I would suggest a limit more like maybe 50,000 downloads?) but I think it would be valuable to protect gems once they are used very widely.

However, if we do add such protection (based on popularity), then we will likely also need to implement a different mechanism to handle security problems with very popular gems. This likely means adding a new command (different from yank) that leaves the gem available, but adds metadata about the security vulnerability, the CVE number, the version with the fix, and something to show those messages to users anytime they install problematic gems.

**Pull request merged**